src/RegistryCompilerPass.php

```diff
@@ -58,23 +73,35 @@ private function buildRegistry(ContainerBuilder $container, $registryId, array $
 					}
 				}
 
-				$collection->add($id, $options);
+				$collection->add([$id, $optionsResolver->getMetadata($options)], $options);
 			}
 
 		}
 
 		$targetId = $registryId;
 		$methodName = $tagOptions['method'];
+		$useMetadata = null !== $tagOptions['metadata'];
+
 		if ($tagOptions['use_collection']) {
 			$targetId = "{$registryId}.collection";
-			$container->setDefinition($targetId, new Definition(\ArrayObject::class))->setPublic(false);
+
+			$collectionClass = $useMetadata ? MetaDataIterator::class : \ArrayObject::class;
+			$container->setDefinition($targetId, new Definition($collectionClass))->setPublic(false);
```

Comment:
yes, for me using `setter` is better than 10 times using an adder, especially if adding one-by-one and sorting is done in realtime vs. using a setter and sorting in compile time - no further footprint introduced in realtime (assuming no logic is written in the setter besides the class attribute assignment)